perfkitbenchmarker/timed_interval.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2014 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A measured interval that generates timing samples."""
+
+from contextlib import contextmanager
+import time
+
+from perfkitbenchmarker import sample
+
+
+class TimedInterval(object):
+  """A measured interval that generates timing samples."""
+
+  def __init__(self, name):
+    """Create a named interval.
+
+    Args:
+      name: A string used to prefix the generated samples' names.
+    """
+    self._name = name
+    self._start_time = None
+    self._stop_time = None
+
+  @contextmanager
+  def Measure(self):
+    """Records the start and stop time of the enclosed interval."""
+    self._start_time = time.time()
+    yield
+    self._stop_time = time.time()
+
+  def GenerateSamples(self, include_runtime=False):
+    """Generates samples based on the times recorded in Measure.
+
+    Args:
+      include_runtime: A Boolean that controls whether an elapsed time
+        sample is included in the generated list.
+
+    Returns:
+      A list of Samples. If Measure has not been called, the list is empty.
+      Otherwise, the list contains at least two Samples. The first Sample
+      contains the timestamp before the enclosed interval, and the second
+      contains the timestamp after the enclosed interval. If include_runtime is
+      True, a third sample is included that contains the elapsed time between
+      the start and stop timestamps.
+    """
+    samples = []
+    if self._stop_time is not None:
+      samples.append(sample.Sample(
+          self._name + ' Start Timestamp', self._start_time, 'seconds'))
+      samples.append(sample.Sample(
+          self._name + ' Stop Timestamp', self._stop_time, 'seconds'))
```

Comment:
I'd lean toward run times only by default (1 sample versus 2)/phase, or omitting these samples entirely unless enabled by a flag.  As it stands now, there's a lot of noise for users just interested in the benchmark result. For example, running Coremark produces one performance sample and 15 timing samples:

```
-------------------------PerfKitBenchmarker Results Summary-------------------------
COREMARK:
        Coremark Score 43544.524276
        Prepare Phase Start Timestamp 1426612739.59 seconds
        Prepare Phase Stop Timestamp 1426612858.99 seconds
        Benchmark Prepare Start Timestamp 1426612834.24 seconds
        Benchmark Prepare Stop Timestamp 1426612858.99 seconds
        Run Phase Start Timestamp 1426612858.99 seconds
        Run Phase Stop Timestamp 1426612899.91 seconds
        Benchmark Run Start Timestamp 1426612858.99 seconds
        Benchmark Run Stop Timestamp 1426612899.91 seconds
        Cleanup Phase Start Timestamp 1426612899.91 seconds
        Cleanup Phase Stop Timestamp 1426612980.97 seconds
        Benchmark Cleanup Start Timestamp 1426612899.91 seconds
        Benchmark Cleanup Stop Timestamp 1426612902.35 seconds
        End to End Start Timestamp 1426612739.59 seconds
        End to End Stop Timestamp 1426612980.97 seconds
        End to End Runtime 241.380285978 seconds
```